{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%matplotlib inline\n",
    "\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "from sklearn.decomposition import PCA \n",
    "from sklearn.ensemble import RandomForestClassifier\n",
    "from sklearn.model_selection import train_test_split\n",
    "from sklearn.metrics import roc_auc_score, roc_curve\n",
    "from sklearn.linear_model import LinearRegression\n",
    "from sklearn.preprocessing import StandardScaler, scale, OneHotEncoder, LabelEncoder, LabelBinarizer\n",
    "import feather\n",
    "import seaborn as sns\n",
    "import matplotlib\n",
    "matplotlib.rcParams['font.sans-serif'] = \"Arial\"\n",
    "matplotlib.rcParams['font.family'] = \"sans-serif\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mean_abundance</th>\n",
       "      <th>conventional</th>\n",
       "      <th>customized</th>\n",
       "      <th>kallisto</th>\n",
       "      <th>salmon</th>\n",
       "      <th>id</th>\n",
       "      <th>samplename</th>\n",
       "      <th>slope</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1.270928</td>\n",
       "      <td>0.139884</td>\n",
       "      <td>-0.115193</td>\n",
       "      <td>1.728354</td>\n",
       "      <td>2.967078</td>\n",
       "      <td>18S_rRNA</td>\n",
       "      <td>Sample_A</td>\n",
       "      <td>8.333992</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>3.563050</td>\n",
       "      <td>2.890183</td>\n",
       "      <td>2.569605</td>\n",
       "      <td>3.573951</td>\n",
       "      <td>4.066626</td>\n",
       "      <td>28S_rRNA</td>\n",
       "      <td>Sample_A</td>\n",
       "      <td>1.681386</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1.874946</td>\n",
       "      <td>0.092692</td>\n",
       "      <td>0.015258</td>\n",
       "      <td>3.031062</td>\n",
       "      <td>3.811763</td>\n",
       "      <td>5.8S_rRNA</td>\n",
       "      <td>Sample_A</td>\n",
       "      <td>17.227866</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1.233572</td>\n",
       "      <td>1.494657</td>\n",
       "      <td>-0.011616</td>\n",
       "      <td>0.676433</td>\n",
       "      <td>2.405615</td>\n",
       "      <td>5S_rRNA</td>\n",
       "      <td>Sample_A</td>\n",
       "      <td>9.662694</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.603084</td>\n",
       "      <td>0.466424</td>\n",
       "      <td>0.401762</td>\n",
       "      <td>0.639327</td>\n",
       "      <td>0.710878</td>\n",
       "      <td>ENSG00000000003</td>\n",
       "      <td>Sample_A</td>\n",
       "      <td>3.998072</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   mean_abundance  conventional  customized  kallisto    salmon  \\\n",
       "0        1.270928      0.139884   -0.115193  1.728354  2.967078   \n",
       "1        3.563050      2.890183    2.569605  3.573951  4.066626   \n",
       "2        1.874946      0.092692    0.015258  3.031062  3.811763   \n",
       "3        1.233572      1.494657   -0.011616  0.676433  2.405615   \n",
       "4        0.603084      0.466424    0.401762  0.639327  0.710878   \n",
       "\n",
       "                id samplename      slope  \n",
       "0         18S_rRNA   Sample_A   8.333992  \n",
       "1         28S_rRNA   Sample_A   1.681386  \n",
       "2        5.8S_rRNA   Sample_A  17.227866  \n",
       "3          5S_rRNA   Sample_A   9.662694  \n",
       "4  ENSG00000000003   Sample_A   3.998072  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_table = '/stor/work/Lambowitz/cdw2854/bench_marking/DEgenes/tpm_table.feather'\n",
    "df = feather.read_dataframe(data_table)  \\\n",
    "    .assign(slope = lambda d: d.salmon / d.customized)     \n",
    "ssc = StandardScaler()\n",
    "X = df.iloc[:,2:-1]\n",
    "scale_tpm = ssc.fit_transform(X)\n",
    "df = pd.DataFrame(scale_tpm, columns=X.columns) \\\n",
    "    .assign(id = df.id) \\\n",
    "    .assign(samplename = df.samplename) \\\n",
    "    .assign(slope = np.abs(df.slope)) \\\n",
    "    .query('samplename == \"Sample_A\"') \\\n",
    "    .reset_index() \\\n",
    "    .drop('index', axis=1) \n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/stor/work/Lambowitz/cdw2854/src/miniconda2/lib/python2.7/site-packages/sklearn/utils/validation.py:429: DataConversionWarning: Data with input dtype int64 was converted to float64 by the scale function.\n",
      "  warnings.warn(msg, _DataConversionWarning)\n"
     ]
    }
   ],
   "source": [
    "df['label'] = map(lambda x,y: 'Bad' if y > x + 1 or y < x - 1 else 'Good', df.salmon, df.customized)\n",
    "gene_length = pd.read_table('/stor/work/Lambowitz/ref/benchmarking/human_transcriptome/genes.length') \\\n",
    "    .assign(gene_length = lambda d: scale(d.gene_length))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['Protein coding', 'Other ncRNA', 'Antisense', 'Other sncRNA',\n",
       "       'rRNA', 'vaultRNA', 'Mt', 'snoRNA', 'miRNA', 'ERCC', 'tRNA'], dtype=object)"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "transcript =  pd.read_table('/stor/work/Lambowitz/ref/benchmarking/human_transcriptome/all_genes.tsv') \\\n",
    "    .rename(columns = {'gene_id':'id'}) \\\n",
    "    .pipe(lambda d: d[['id', 'type']]) \\\n",
    "    .query('type!=\"LRG_gene\"')\n",
    "le = LabelEncoder()\n",
    "ohe = OneHotEncoder(sparse=False)\n",
    "le.fit(transcript['type'])\n",
    "type_code = le.transform(transcript['type'])\n",
    "transcript.type.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mean_abundance</th>\n",
       "      <th>conventional</th>\n",
       "      <th>customized</th>\n",
       "      <th>kallisto</th>\n",
       "      <th>salmon</th>\n",
       "      <th>id</th>\n",
       "      <th>samplename</th>\n",
       "      <th>slope</th>\n",
       "      <th>label</th>\n",
       "      <th>gene_length</th>\n",
       "      <th>...</th>\n",
       "      <th>ERCC</th>\n",
       "      <th>Mt</th>\n",
       "      <th>Other ncRNA</th>\n",
       "      <th>Other sncRNA</th>\n",
       "      <th>Protein coding</th>\n",
       "      <th>miRNA</th>\n",
       "      <th>rRNA</th>\n",
       "      <th>snoRNA</th>\n",
       "      <th>tRNA</th>\n",
       "      <th>vaultRNA</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1.270928</td>\n",
       "      <td>0.139884</td>\n",
       "      <td>-0.115193</td>\n",
       "      <td>1.728354</td>\n",
       "      <td>2.967078</td>\n",
       "      <td>18S_rRNA</td>\n",
       "      <td>Sample_A</td>\n",
       "      <td>8.333992</td>\n",
       "      <td>Bad</td>\n",
       "      <td>-0.314249</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>3.563050</td>\n",
       "      <td>2.890183</td>\n",
       "      <td>2.569605</td>\n",
       "      <td>3.573951</td>\n",
       "      <td>4.066626</td>\n",
       "      <td>28S_rRNA</td>\n",
       "      <td>Sample_A</td>\n",
       "      <td>1.681386</td>\n",
       "      <td>Bad</td>\n",
       "      <td>-0.276548</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1.874946</td>\n",
       "      <td>0.092692</td>\n",
       "      <td>0.015258</td>\n",
       "      <td>3.031062</td>\n",
       "      <td>3.811763</td>\n",
       "      <td>5.8S_rRNA</td>\n",
       "      <td>Sample_A</td>\n",
       "      <td>17.227866</td>\n",
       "      <td>Bad</td>\n",
       "      <td>-0.334673</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1.233572</td>\n",
       "      <td>1.494657</td>\n",
       "      <td>-0.011616</td>\n",
       "      <td>0.676433</td>\n",
       "      <td>2.405615</td>\n",
       "      <td>5S_rRNA</td>\n",
       "      <td>Sample_A</td>\n",
       "      <td>9.662694</td>\n",
       "      <td>Bad</td>\n",
       "      <td>-0.335102</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.603084</td>\n",
       "      <td>0.466424</td>\n",
       "      <td>0.401762</td>\n",
       "      <td>0.639327</td>\n",
       "      <td>0.710878</td>\n",
       "      <td>ENSG00000000003</td>\n",
       "      <td>Sample_A</td>\n",
       "      <td>3.998072</td>\n",
       "      <td>Good</td>\n",
       "      <td>-0.183032</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   mean_abundance  conventional  customized  kallisto    salmon  \\\n",
       "0        1.270928      0.139884   -0.115193  1.728354  2.967078   \n",
       "1        3.563050      2.890183    2.569605  3.573951  4.066626   \n",
       "2        1.874946      0.092692    0.015258  3.031062  3.811763   \n",
       "3        1.233572      1.494657   -0.011616  0.676433  2.405615   \n",
       "4        0.603084      0.466424    0.401762  0.639327  0.710878   \n",
       "\n",
       "                id samplename      slope label  gene_length    ...     ERCC  \\\n",
       "0         18S_rRNA   Sample_A   8.333992   Bad    -0.314249    ...      0.0   \n",
       "1         28S_rRNA   Sample_A   1.681386   Bad    -0.276548    ...      0.0   \n",
       "2        5.8S_rRNA   Sample_A  17.227866   Bad    -0.334673    ...      0.0   \n",
       "3          5S_rRNA   Sample_A   9.662694   Bad    -0.335102    ...      0.0   \n",
       "4  ENSG00000000003   Sample_A   3.998072  Good    -0.183032    ...      0.0   \n",
       "\n",
       "    Mt  Other ncRNA  Other sncRNA  Protein coding  miRNA  rRNA  snoRNA  tRNA  \\\n",
       "0  0.0          0.0           0.0             0.0    0.0   1.0     0.0   0.0   \n",
       "1  0.0          0.0           0.0             0.0    0.0   1.0     0.0   0.0   \n",
       "2  0.0          0.0           0.0             0.0    0.0   1.0     0.0   0.0   \n",
       "3  0.0          0.0           0.0             0.0    0.0   1.0     0.0   0.0   \n",
       "4  0.0          0.0           0.0             1.0    0.0   0.0     0.0   0.0   \n",
       "\n",
       "   vaultRNA  \n",
       "0       0.0  \n",
       "1       0.0  \n",
       "2       0.0  \n",
       "3       0.0  \n",
       "4       0.0  \n",
       "\n",
       "[5 rows x 21 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "X = ohe.fit_transform(np.array([type_code, type_code]).transpose())\n",
    "ncol = X.shape[1]\n",
    "type_X = pd.DataFrame(X[:,:ncol/2], columns = le.classes_)\n",
    "type_X['id'] = transcript['id']\n",
    "df = df.merge(gene_length).merge(type_X)\n",
    "df.head()\n",
    "columns = df.columns[df.columns.str.contains('conventional|customized|kallisto|salmon|id|samplename|label|slope')]\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def plot_gene_expression(df, ax):\n",
    "    X = df['salmon']\n",
    "    Y = df['customized']\n",
    "    ax.scatter(X, Y)\n",
    "    x = np.array([-10, 10])\n",
    "    ax.plot(x, x, color='red')\n",
    "    ax.plot(x, x+1 , color='green')\n",
    "    ax.plot(x, x-1, color='green')\n",
    "    ax.set_xlim(0,10)\n",
    "    ax.set_ylim(0,10)\n",
    "    ax.spines['top'].set_visible(False)\n",
    "    ax.spines['right'].set_visible(False)\n",
    "    ax.set_xlabel('Salmon')\n",
    "    ax.set_ylabel('Customized pipeline')\n",
    "    return 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def plot_roc(Y_test, pred_y, ax):\n",
    "    auc = roc_auc_score(Y_test, pred_y[:,1])\n",
    "    fpr, tpr, _ = roc_curve(Y_test, pred_y[:,1])\n",
    "    ax.plot(fpr,tpr)\n",
    "    ax.set_xlabel('False Positive Rate')\n",
    "    ax.set_ylabel('Ttrue Positive Rate')\n",
    "    ax.text(0.6,0.2, 'AUC: %.3f' %(auc))\n",
    "    ax.plot([0,1],[0,1], color='red')\n",
    "    ax.spines['top'].set_visible(False)\n",
    "    ax.spines['right'].set_visible(False)\n",
    "    return 0\n",
    "\n",
    "def plot_importance(rf, columns, ax):\n",
    "    import_df = pd.DataFrame({'import':rf.feature_importances_,\n",
    "              'feature':columns}) \\\n",
    "        .sort_values('import', ascending=False)\n",
    "    import_df.plot(kind='bar', ax = ax)\n",
    "    ax.legend().set_visible(False)\n",
    "    ax.set_xlabel(' ')\n",
    "    ax.set_ylabel('Importance (%)')\n",
    "    ax.set_xticklabels(import_df.feature, rotation=90)\n",
    "    ax.spines['top'].set_visible(False)\n",
    "    ax.spines['right'].set_visible(False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "could not convert string to float: Good",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m\u001b[0m",
      "\u001b[0;31mValueError\u001b[0mTraceback (most recent call last)",
      "\u001b[0;32m<ipython-input-8-b091fb524902>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m     10\u001b[0m     \u001b[0;32mreturn\u001b[0m \u001b[0mrf\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mpred_y\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mY_test\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     11\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 12\u001b[0;31m \u001b[0mrf\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mpred_y\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mY_test\u001b[0m \u001b[0;34m=\u001b[0m  \u001b[0mtrain_rf\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdf\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcolumns\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m<ipython-input-8-b091fb524902>\u001b[0m in \u001b[0;36mtrain_rf\u001b[0;34m(df, columns)\u001b[0m\n\u001b[1;32m      6\u001b[0m     \u001b[0mrf\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mRandomForestClassifier\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mn_jobs\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m-\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      7\u001b[0m     \u001b[0mX_train\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mX_test\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mY_train\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mY_test\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mtrain_test_split\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mX\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mY\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtest_size\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m0.2\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 8\u001b[0;31m     \u001b[0mrf\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfit\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mX_train\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mY_train\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      9\u001b[0m     \u001b[0mpred_y\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mrf\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpredict_proba\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mX_test\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     10\u001b[0m     \u001b[0;32mreturn\u001b[0m \u001b[0mrf\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mpred_y\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mY_test\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/stor/work/Lambowitz/cdw2854/src/miniconda2/lib/python2.7/site-packages/sklearn/ensemble/forest.pyc\u001b[0m in \u001b[0;36mfit\u001b[0;34m(self, X, y, sample_weight)\u001b[0m\n\u001b[1;32m    245\u001b[0m         \"\"\"\n\u001b[1;32m    246\u001b[0m         \u001b[0;31m# Validate or convert input data\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 247\u001b[0;31m         \u001b[0mX\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mcheck_array\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mX\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0maccept_sparse\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m\"csc\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdtype\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mDTYPE\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    248\u001b[0m         \u001b[0my\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mcheck_array\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0my\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0maccept_sparse\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'csc'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mensure_2d\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mFalse\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdtype\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mNone\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    249\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0missparse\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mX\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/stor/work/Lambowitz/cdw2854/src/miniconda2/lib/python2.7/site-packages/sklearn/utils/validation.pyc\u001b[0m in \u001b[0;36mcheck_array\u001b[0;34m(array, accept_sparse, dtype, order, copy, force_all_finite, ensure_2d, allow_nd, ensure_min_samples, ensure_min_features, warn_on_dtype, estimator)\u001b[0m\n\u001b[1;32m    380\u001b[0m                                       force_all_finite)\n\u001b[1;32m    381\u001b[0m     \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 382\u001b[0;31m         \u001b[0marray\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0marray\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0marray\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdtype\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mdtype\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0morder\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0morder\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcopy\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mcopy\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    383\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    384\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mensure_2d\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mValueError\u001b[0m: could not convert string to float: Good"
     ]
    }
   ],
   "source": [
    "def train_rf(df, columns):\n",
    "    X = df[columns]\n",
    "    Y = df['label']\n",
    "    lb = LabelBinarizer()\n",
    "    Y = lb.fit_transform(Y)\n",
    "    rf = RandomForestClassifier(n_jobs=-1)\n",
    "    X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.2)\n",
    "    rf.fit(X_train, Y_train)\n",
    "    pred_y = rf.predict_proba(X_test)\n",
    "    return rf, pred_y, Y_test\n",
    "\n",
    "rf, pred_y, Y_test =  train_rf(df, columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "figurepath = '/stor/work/Lambowitz/cdw2854/bench_marking/figures'\n",
    "figurename = figurepath + '/roc_discordant.png'\n",
    "\n",
    "FONT_SIZE = 20\n",
    "\n",
    "plt.rc('font', size=FONT_SIZE)          # controls default text sizes\n",
    "plt.rc('axes', titlesize=FONT_SIZE)     # fontsize of the axes title\n",
    "plt.rc('axes', labelsize=FONT_SIZE)    # fontsize of the x and y labels\n",
    "plt.rc('xtick', labelsize=FONT_SIZE)    # fontsize of the tick labels\n",
    "plt.rc('ytick', labelsize=FONT_SIZE)    # fontsize of the tick labels\n",
    "plt.rc('legend', fontsize=FONT_SIZE)    # legend fontsize\n",
    "plt.rc('figure', titlesize=FONT_SIZE)\n",
    "\n",
    "sns.set_style('white')\n",
    "fig = plt.figure(figsize=(10,10))\n",
    "ax_gene = fig.add_axes([0.08,0.5,1,0.5])\n",
    "ax_roc = fig.add_axes([0.08,0,0.3,0.4])\n",
    "ax_importance = fig.add_axes([0.5, 0,0.5,0.4])\n",
    "plot_gene_expression(df, ax_gene)\n",
    "plot_roc(Y_test, pred_y, ax_roc)\n",
    "plot_importance(rf, columns, ax_importance)\n",
    "fig.text(0,1,'a', weight='bold', fontsize=25)\n",
    "fig.text(0,0.4,'b',weight='bold', fontsize=25)\n",
    "fig.text(0.45,0.4,'c',weight='bold', fontsize=25)\n",
    "plt.savefig(figurename, transparent=True, bbox_inches='tight')\n",
    "print 'Saved: ', figurename"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "miniconda2",
   "language": "python",
   "name": "miniconda2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
